### PR TITLE
improvement: count to for_each

### LINKED ISSUE #1105: Feat/bug: workers ASGs that are more resilient to unnecessary change
# I have issues

## I'm submitting a...

* [X] bug report
* [X] feature request
* [ ] support request - read the [FAQ](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/faq.md) first!
* [X] kudos, thank you, warm fuzzy

## What is the current behavior?

If I mutate the array I pass in for `worker_groups` at all, all of my ASGs are regenerated and it's kinda explosive

## If this is a bug, how to reproduce? Please include a code sample if relevant.

Pass in a list for `worker_groups`, append an entry on a subsequent run, note the resources to destroy, try not to cry, cry anyway.

## What's the expected behavior?

By preference, only ASGs that didn't exist before would be added, or if existing ASGs were changed would they be replaced.

## Are you able to fix this problem and submit a PR? Link here if you have already.

Not really but I had an idea that instead of a count loop (https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/workers.tf#L4 ) we could use a for_each loop, maybe. A map would be better for this than a list, so perhaps a different ASG fed by a new module param (`worker_groups_map`?) would be needed for backwards compatibility. Or maybe you could generate a map from a list I'm no scientist.

## Environment details

* Affected module version: `12.2.0` but the code looks the same in `13.2.1`
* OS: Linux
* Terraform version: `0.13.4`

## Any other relevant info

Thanks for this module, I've been using it I think actual years plural now, and it's always super useful and composable (in fact, for this ticket I could easily create my own `workers.tf` and have it work with your existing module easy peasy)

### PR DESCRIPTION
**Pull request by sbeginCoveo**
# PR o'clock

## Description

Closes #1105 
Partially addresses #635 

There are two themes to this PR. 
- One is merging the default values in the locals to avoid unnecessary lookups, which pollutes the codebase
- The main theme is migrating `count` to `for_each`, so that we can use relevant key names for identification
```
module.terraform_aws_eks.aws_autoscaling_group.workers["node-1"]
module.terraform_aws_eks.aws_autoscaling_group.workers["node-spots-2"]
module.terraform_aws_eks.aws_autoscaling_group.workers["gpu-1"]
...
module.terraform_aws_eks.aws_launch_configuration.workers["node-1"]
module.terraform_aws_eks.aws_launch_configuration.workers["node-spots-2"]
module.terraform_aws_eks.aws_launch_configuration.workers["gpu-1"]
```

## State migration process

There are no breaking changes in the interface (variables, outputs), But we have a tfstate migration required to move to the for_each index pattern.

The four main resources to migrate are:
- `aws_autoscaling_group.workers`
- `aws_iam_instance_profile.workers`
- `aws_launch_configuration.workers`
- `random_pet.workers`

> This can be done a variety of ways, but Ive built a migration-file-generator which feeds on a terraform plan output and prints `terraform state mv` commands. Please tell me if you are interested, and I'll look into open sourcing it. 

Example migration
```shell
terraform state mv module.terraform_aws_eks.aws_autoscaling_group.workers[0]  module.terraform_aws_eks.aws_autoscaling_group.workers["node-0"]
terraform state mv module.terraform_aws_eks.aws_autoscaling_group.workers[1]  module.terraform_aws_eks.aws_autoscaling_group.workers["node-1"]
terraform state mv module.terraform_aws_eks.aws_autoscaling_group.workers[2]  module.terraform_aws_eks.aws_autoscaling_group.workers["node-spots-0"]

terraform state mv module.terraform_aws_eks.aws_iam_instance_profile.workers[0]  module.terraform_aws_eks.aws_iam_instance_profile.workers["node-0"]
terraform state mv module.terraform_aws_eks.aws_iam_instance_profile.workers[1]  module.terraform_aws_eks.aws_iam_instance_profile.workers["node-1"]
terraform state mv module.terraform_aws_eks.aws_iam_instance_profile.workers[2]  module.terraform_aws_eks.aws_iam_instance_profile.workers["node-spots-0"]

terraform state mv module.terraform_aws_eks.aws_launch_configuration.workers[0]  module.terraform_aws_eks.aws_launch_configuration.workers["node-0"]
terraform state mv module.terraform_aws_eks.aws_launch_configuration.workers[1]  module.terraform_aws_eks.aws_launch_configuration.workers["node-1"]
terraform state mv module.terraform_aws_eks.aws_launch_configuration.workers[2]  module.terraform_aws_eks.aws_launch_configuration.workers["node-spots-0"]

terraform state mv module.terraform_aws_eks.random_pet.workers[0]  module.terraform_aws_eks.random_pet.workers["node-0"]
terraform state mv module.terraform_aws_eks.random_pet.workers[1]  module.terraform_aws_eks.random_pet.workers["node-1"]
terraform state mv module.terraform_aws_eks.random_pet.workers[2]  module.terraform_aws_eks.random_pet.workers["node-spots-0"]
```

## Checklist

- [x] CI tests are passing
- [x] README.md has been updated after any changes to variables and outputs. See https://github.com/terraform-aws-modules/terraform-aws-eks/#doc-generation (no interface change)

### CODE REVIEW

**Review by james-callahan — APPROVED** [review]
I think this is a great first step to getting #858 integrated.

The one thing I'll comment on is: as the `for_each` key is not customizable, we still have the  issues with re-ordering as worker groups are added/removed, which will require another state migration to fix.

File: local.tf
Comment (james-callahan) [inline local.tf:116]: Actually, would this work?:
```suggestion
    for key, template in var.worker_groups_launch_template : lookup(template, "name", key) => merge(
```

File: outputs.tf
Comment (james-callahan) [inline outputs.tf:87]: I'm not sure if order will be retained here?

File: local.tf
Comment (sbeginCoveo) [reply local.tf:116]: :heavy_check_mark:

File: outputs.tf
Comment (sbeginCoveo) [reply outputs.tf:87]: From what I can gather, I think so, because each set of workers will now be sorted by name, instead of its original input order, from the variables provided by the user.

> _values are returned in lexicographical order by their corresponding keys_
> https://www.terraform.io/docs/language/functions/values.html

Here's a few of my thoughts on possible ways to solve that
- If we want to preserve that order at all costs, we could save the original order as a key in our objects, then sort using this key at the output level.
- Since were introducing a potential breaking change in the usage of this output, we could take the time to change them to maps, with their respective names as keys, or other strategy.
- If we want to minimise impact, we could leave it like this so that only the order is unpreserved, and make the necessary announcements in the release/migration process

Comment (james-callahan) [reply outputs.tf:87]: > If we want to preserve that order at all costs, we could save the original order as a key in our objects, then sort using this key at the output level.

I think we might need to, to make this an easy merge. 

> Since were introducing a potential breaking change in the usage of this output, we could take the time to change them to maps, with their respective names as keys, or other strategy.

Agreed, and even if we do need to keep the old array in order then we should add this as a separate output.

**Comment by james-callahan** [comment]
Look like this needs rebasing now

File: outputs.tf
Comment (sbeginCoveo) [reply outputs.tf]: I chose to preserve the ordering to make this _an easy merge_ like you said. more outputs can be added later :)

File: aws_auth.tf
Comment (james-callahan) [inline aws_auth.tf:5]: nit pick: should there be whitespace after `in`?

File: local.tf
Comment (james-callahan) [inline local.tf:157]: Could these be e.g.:
```
[ for g in var.worker_groups : aws_autoscaling_group.worker[g.name] ]
```

File: workers.tf
Comment (james-callahan) [inline workers.tf:80]: Do we need something similar to ffdc7a79a575d0c1866bc97d3d7c804990cd66a9 here?

**Review by james-callahan — APPROVED** [review]

File: aws_auth.tf
Comment (sbeginCoveo) [reply aws_auth.tf:5]: `terraform fmt` removes them :/

File: local.tf
Comment (sbeginCoveo) [reply local.tf]: youre right. with the appropriate conditions at first this works and is a bit cleaner

File: workers.tf
Comment (sbeginCoveo) [reply workers.tf]: :heavy_check_mark:

File: local.tf
Comment (james-callahan) [inline local.tf:124]: I get an error here:
>This object does not have an attribute named "additional_ebs_volumes".

**Review by james-callahan — CHANGES_REQUESTED** [review]

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:163]: This fails with:
> The given key does not identify an element in this collection value.

i.e. when `manage_worker_iam_resourcesmanage_worker_iam_resources` is true it's an empty collection.

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:164]:
```suggestion
    name = coalesce(
      aws_iam_instance_profile.workers_launch_template,
      data.aws_iam_instance_profile.custom_worker_group_launch_template_iam_instance_profile,
    )[each.key].name
```

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:168]: This fails. did you mean:
```suggestion
    enabled = each.value.enclave_support
```

File: local.tf
Comment (james-callahan) [inline local.tf:120]: if you put `{ name => key }` in here then you could avoid `lookup(each.value, "name", each.key)` elsewhere

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:37]: I think this (and many other similar lines) need to use `each.value` rather than `var.worker_groups_launch_template[each.key]`

```suggestion
    for_each = (lookup(each.value, "override_instance_types", null) != null) || (each.value.on_demand_allocation_strategy != null) ? list(each.value) : []
```

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:91]:
```suggestion
    for_each = lookup(each.value, "warm_pool", null) != null ? [lookup(each.value, "warm_pool")] : []
```

File: local.tf
Comment (sbeginCoveo) [reply local.tf:124]: fixed :heavy_check_mark:

File: workers_launch_template.tf
Comment (sbeginCoveo) [reply workers_launch_template.tf:168]: changed :heavy_check_mark:

File: workers_launch_template.tf
Comment (sbeginCoveo) [reply workers_launch_template.tf:163]: fixed with your suggestion :heavy_check_mark:

File: workers_launch_template.tf
Comment (sbeginCoveo) [reply workers_launch_template.tf:164]: fixed with your suggestion :heavy_check_mark:

File: workers_launch_template.tf
Comment (sbeginCoveo) [reply workers_launch_template.tf:91]: fixed multiple occurrences of the same pattern :heavy_check_mark:

File: local.tf
Comment (sbeginCoveo) [reply local.tf:120]: I added these changes :heavy_check_mark:

**Comment by sbeginCoveo** [comment]
@james-callahan 
> The one thing I'll comment on is: as the `for_each` key is not customizable, we still have the issues with re-ordering as worker groups are added/removed, which will require another state migration to fix.

Ive played with the idea of adding a dedicated identifier that would be used as an index for maps, but the name is still used in `name_prefix`, forcing an ASG recreation, therefore requiring to be unique and unchanging as would the id. This forces clients to make their names/id unique and unchanging over time in order to profit from the resiliency introduced by `for_each`.

So at this point I don't see an advantage to adding an additional parameter. This set of changes can be merged _as is_ and clients will have to replace their ASG by changing their names if they want to profit from the order preservation feature offered by for_each.

Tell me what you think :)

File: local.tf
Comment (james-callahan) [inline local.tf:91]: This change causes `dynamic "mixed_instances_policy"`'s `for_each` to activate; resulting in unconditionally enabling mixed instances.

**Review by james-callahan — CHANGES_REQUESTED** [review]

File: local.tf
Comment (sbeginCoveo) [reply local.tf]: fixed :heavy_check_mark:

**Comment by sbeginCoveo** [comment]
:wave: were are we at regarding this feature? Its required to enable us to have unique keys on autoscaling groups

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:36]:
```
workers_launch_template.tf line 36, in resource "aws_autoscaling_group" "workers_launch_template":
│   36:     for_each = (lookup(each.value, "override_instance_types", null) != null) || (each.value.on_demand_allocation_strategy != null) ? list(each.value) : []
│     ├────────────────
│     │ each.value is object with 70 attributes
│ 
│ Call to function "list" failed: the "list" function was deprecated in Terraform v0.12 and is no longer available; use
│ tolist([ ... ]) syntax to write a literal list.
```

**Review by james-callahan — CHANGES_REQUESTED** [review]

File: workers_launch_template.tf
Comment (james-callahan) [reply workers_launch_template.tf:36]:
```suggestion
    for_each = (lookup(each.value, "override_instance_types", null) != null) || (each.value.on_demand_allocation_strategy != null) ? [each.value] : []
```

File: workers_launch_template.tf
Comment (james-callahan) [inline workers_launch_template.tf:68]:
```suggestion
    for_each = (lookup(each.value, "override_instance_types", null) != null) || (each.value.on_demand_allocation_strategy != null) ? [] : [each.value]
```

File: workers_launch_template.tf
Comment (sbeginCoveo) [reply workers_launch_template.tf]: :heavy_check_mark:

File: workers_launch_template.tf
Comment (sbeginCoveo) [reply workers_launch_template.tf]: :heavy_check_mark:

**Review by james-callahan — APPROVED** [review]
LGTM; would love to see this merged now.


Only comment I'll add is that the commits are a bit of a mess; I'm not sure if this repository does squash merging or what the story is.

**Comment by sbeginCoveo** [comment]
want me to force-push in a single commit?

**Comment by barryib** [comment]
Hello @sbeginCoveo Thank you for your work.

We do plan to move from count to for_each. But to that, we planned to move code into sub-module and for_each. I can't merge this because it doesn't go into that direction. I'll keep this PR open for now, your work idea of merging locals is great and will contact you soon to add this in a `next` branch.

**Comment by james-callahan** [comment]
> But to that, we planned to move code into sub-module and for_each.

What do you mean by that?

> I can't merge this because it doesn't go into that direction

Isn't this already an improvement? could the move to a submodule happen as a second step?

----

@sbeginCoveo seems the PR needs rebasing after another recent merge

**Comment by barryib** [comment]
> > But to that, we planned to move code into sub-module and for_each.
> 
> What do you mean by that?


I mean that we will move code into submodules and use for_each to loop over them.

Please see

https://github.com/terraform-aws-modules/terraform-aws-eks/issues/774#issuecomment-663165563
https://github.com/terraform-aws-modules/terraform-aws-eks/issues/635
https://github.com/terraform-aws-modules/terraform-aws-eks/issues/1101

**Comment by sbeginCoveo** [comment]
well, I probably won't be maintaining this branch since we don't intend to merge it.
feel free to reuse the local merging strategy in another PR :wave:

**Comment by innovia** [comment]
@barryib When is this going to happen? What's the workaround - I currently have a huge mess when I need to remove and asg from the list.

I thought about decoupling this into a module like you said, but if you have any updates or pointers please let me know

**Pull request closed without merging**
